library/src/main/camera2/com/google/android/cameraview/Camera2.java

```
@@ -602,40 +629,85 @@ private void collectCameraInfo() {
                 mPreviewSizes.add(new Size(width, height));
             }
         }
-        mPictureSizes.clear();
-        collectPictureSizes(map);
+        mOutputSizes.clear();
+        collectOutputSizes(map);
         for (AspectRatio ratio : mPreviewSizes.ratios()) {
-            if (!mPictureSizes.ratios().contains(ratio)) {
+            if (!mOutputSizes.ratios().contains(ratio)) {
                 mPreviewSizes.remove(ratio);
             }
         }
 
-        if (!mPreviewSizes.ratios().contains(mAspectRatio)) {
-            mAspectRatio = mPreviewSizes.ratios().iterator().next();
+        mAspectRatio = chooseBestAspectRatio();
+    }
+
+    /**
+     * @return the best supported aspect ratio based on the preferred ratios.
+     * If none of the preferred ratios are available, it chooses the ratio of the largest preview
+     * size that matches the screen orientation.
+     * It returns null if the list of preview sizes hasn't been collected yet
+     */
+    @Nullable
+    private AspectRatio chooseBestAspectRatio() {
+        if (mPreviewSizes == null || mPreviewSizes.ratios().isEmpty()) {
+            return null;
```

should it throw rather than return null? Unsure if it's likely 🤔

---

It will return null very often, for example when we set the preferred ratios but the camera is not opened yet so we cannot check whether they are supported or not.